{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e2ff4350-29b3-4043-89c9-f2fc28b207d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import itertools as it\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import math\n",
    "import re\n",
    "from sklearn.preprocessing import normalize\n",
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "10027741-feec-4d35-b112-af613441a111",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Frequency(object):\n",
    "    \n",
    "    def __init__(self, k = 7, split = None, segment = None, quality = None, qualpos = 0, segpos = 0, variable = 0.9):\n",
    "    \n",
    "        self.k = k\n",
    "        self.qualpos = qualpos\n",
    "        self.segpos = segpos\n",
    "        self.split = split\n",
    "        self.segment = segment\n",
    "        self.quality = quality\n",
    "        self.variable = variable\n",
    "        \n",
    "        self.nucleotides = ['A', 'C', 'G', 'T']\n",
    "        self.substit = dict.fromkeys(map(ord, self.nucleotides), None)\n",
    "        self.exist = dict.fromkeys(map(''.join, it.product(self.nucleotides, repeat = self.k)), 0)        \n",
    "        self.col = len(self.exist.keys())\n",
    "\n",
    "        self.nucex = {\n",
    "            'A':['A'],\n",
    "            'C':['C'],\n",
    "            'G':['G'],\n",
    "            'T':['T'],\n",
    "            'R':['A', 'G'],\n",
    "            'Y':['C', 'T'],\n",
    "            'W':['A', 'T'],\n",
    "            'S':['C', 'G'],\n",
    "            'M':['A', 'C'],\n",
    "            'K':['G', 'T'],\n",
    "            'B':['G', 'C', 'T'],\n",
    "            'H':['A', 'C', 'T'],\n",
    "            'D':['A', 'G', 'T'],\n",
    "            'V':['A', 'C', 'G'],\n",
    "            'N':['A', 'C', 'G', 'T'],\n",
    "        } \n",
    "        self.nucmut = {\n",
    "            'A':['C', 'G', 'T'],\n",
    "            'C':['A', 'G', 'T'],\n",
    "            'G':['A', 'C', 'T'],\n",
    "            'T':['A', 'C', 'G'],\n",
    "        } \n",
    "    \n",
    "    def checkup(self, name):\n",
    "        \n",
    "        try:\n",
    "            if self.segment and self.quality:    \n",
    "                head = name.split(self.split)\n",
    "                if re.match(self.segment, head[self.segpos], re.IGNORECASE) and re.match(self.quality, head[self.qualpos], re.IGNORECASE):\n",
    "                    return(True)\n",
    "                else:\n",
    "                    return(False)\n",
    "\n",
    "            elif self.segment or self.quality:\n",
    "                head = name.split(self.split)\n",
    "                \n",
    "                if e.match(self.segment, head[self.segpos], re.IGNORECASE) or re.match(self.quality, head[self.qualpos], re.IGNORECASE):\n",
    "                    return(True)\n",
    "                else:\n",
    "                    return(False)\n",
    "\n",
    "            else:\n",
    "                return(True)\n",
    "        except:\n",
    "            return(False)\n",
    "            \n",
    "    def countRows(self, infile):\n",
    "        \n",
    "        row = 0\n",
    "        for entry in SeqIO.parse(infile,'fasta'):\n",
    "            \n",
    "            name = entry.name\n",
    "            sequence = str(entry.seq)\n",
    "            missing = len(sequence.translate(self.substit))\n",
    "            fracture = float(len(sequence)/missing) if missing else 0 \n",
    "            \n",
    "            if self.checkup(name) == True and fracture <= self.variable:\n",
    "                row += 1\n",
    "                \n",
    "        return(row)\n",
    "    \n",
    "    def calculateFrequence(self, infile):\n",
    "        \n",
    "        row = self.countRows(infile)\n",
    "        index = np.empty(row, dtype = 'object')\n",
    "        matrix = np.empty((row, self.col, ),dtype = 'float32')\n",
    "        \n",
    "        pos = 0\n",
    "        for entry in SeqIO.parse(infile,'fasta'):\n",
    "            \n",
    "            name = entry.name\n",
    "            sequence = str(entry.seq)\n",
    "            accession = name.split(self.split)[0]\n",
    "            missing = len(sequence.translate(self.substit))\n",
    "            fracture = float(len(sequence)/missing) if missing else 0 \n",
    "            \n",
    "            if self.checkup(name) == True and fracture <= self.variable:\n",
    "                for i in range(len(sequence) - self.k + 1):\n",
    "                    \n",
    "                    kmer = sequence[i:i+self.k]\n",
    "                    \n",
    "                    if fracture == 0:\n",
    "                        main = [kmer]\n",
    "                        size = 1\n",
    "                    else:\n",
    "                        main = map(''.join, it.product(*[self.nucex.get(j) for j in kmer]))\n",
    "                        size = np.prod([len(self.nucex.get(k)) for k in kmer])\n",
    "                        \n",
    "                    for sub in main:\n",
    "                        self.exist[sub] += float(1/size)\n",
    "                    \n",
    "                    #     for l, nuc in enumerate(main):\n",
    "                    #         self.exist[main] += (1-self.mutfac)/size\n",
    "                    #         for mutation in map(''.join, it.product(*[[main[:l]], nucmut.get(nuc), [main[l+1:]]])):\n",
    "                    #             self.exist[mutation] += self.mutfac/(size*12)\n",
    "\n",
    "                # matrix[pos] = normalize(np.fromiter(self.exist.values(), dtype = '<f8', count = self.col).reshape(1, -1), norm = 'l1')\n",
    "                matrix[pos] = np.fromiter(self.exist.values(), dtype = 'float32', count = self.col)/sum(self.exist.values())\n",
    "                index[pos] = accession\n",
    "                \n",
    "                self.exist.update((k,0) for k in self.exist.keys())\n",
    "                pos += 1\n",
    "            \n",
    "        return(index, matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "abc69bf4-9663-4adc-8c87-1b06779b2003",
   "metadata": {},
   "outputs": [],
   "source": [
    "Freq = Frequency(k = 7, split = '|', segment = '4', quality = 'Pass', qualpos = 8, segpos = 2, variable = 0.9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1239b5ae-3489-4ee1-ad55-3b432e16ef39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%timeit\n",
    "index, matrix = Freq.calculateFrequence('A.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4f710926-1845-4485-86e7-556cf8221d90",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca_init = PCA(n_components=50)\n",
    "pca_matr = pca_init.fit_transform(matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "69d602fa-644b-4616-bb1e-539fba414a9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 6.74045645e-03,  6.15994493e-03,  6.81750290e-03, ...,\n",
       "         1.92624901e-03, -1.01706781e-03, -1.90419995e-03],\n",
       "       [ 6.51352294e-03,  6.00963552e-03,  7.93484226e-03, ...,\n",
       "         1.75852340e-03, -1.29840942e-03, -2.02510972e-03],\n",
       "       [ 6.36859238e-03,  6.09499076e-03,  8.61809310e-03, ...,\n",
       "         2.03404063e-03, -1.14415598e-03, -2.13219831e-03],\n",
       "       ...,\n",
       "       [ 1.65543742e-02, -8.60017259e-03, -4.68872546e-04, ...,\n",
       "         7.43104756e-05,  4.00506724e-05,  2.75053142e-04],\n",
       "       [ 2.22518248e-03,  9.64119472e-03, -4.93186014e-03, ...,\n",
       "         1.14396017e-03,  1.08391956e-04,  2.67903351e-06],\n",
       "       [ 3.19739641e-03,  9.13690124e-03, -3.88645823e-03, ...,\n",
       "         1.16204159e-04, -1.96156805e-04,  2.45552365e-04]], dtype=float32)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pca_matr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb14bbf1-b31d-48e9-abac-0c43e7bf226d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
